lib/elixir/lib/kernel.ex

```diff
@@ -2198,6 +2198,30 @@ defmodule Kernel do
     :erlang.error(ArgumentError.exception(error_message))
   end
 
+  @doc """
+  Returns true if `term` is a structure; otherwise returns `false`.
+
+  Allowed in guard tests.
+  """
+  @doc guard: true
+  defmacro is_struct(term) do
+    case __CALLER__.context do
+      nil ->
+        quote do
+          case unquote(term) do
+            %_{} -> true
+            _ -> false
+          end
+        end
+
+      :match ->
+        invalid_match!(:is_struct)
+
+      :guard ->
+        quote(do: is_map(unquote(term)) and is_atom(:erlang.map_get(:__struct__, unquote(term))))
```

Comment:
We are missing the `:erlang.is_map_key(:__struct__, unquote(term))` check here.

To validate, write this test:

    defp check_struct_or_map(arg) when is_struct(arg) or is_map(arg), do: true
    defp check_struct_or_map(_arg), do: false

It should return true for a map, but it will return false.

Reply:
Done.